### PR TITLE
Now supporting line drawing to native YUV

### PR DESCRIPTION
**Pull request by emilwestergren**
@davidpiuva

### CODE REVIEW

File: source/draw/gpu/opengl/shaders/whiteToNativeYUV.frag [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#version 300 es
+#extension GL_EXT_YUV_target : enable
+precision highp float;
+uniform vec4 color;
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/shaders/whiteToNativeYUV.frag]: Will color be used in another pull request?

Comment (davidpiuva) [inline source/draw/gpu/opengl/shaders/whiteToNativeYUV.frag]: You can also try to make it work in the same PR since it is a small thing.

Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/whiteToNativeYUV.frag]: Sure

File: source/draw/gpu/opengl/OpenGLContext.ooc [hunk 2]
```diff
@@ -81,11 +81,14 @@ OpenGLContext: class extends GpuContext {
...
 	}
+	getLineShader: virtual func -> OpenGLMap { this _linesShader }
+	getDefaultShader: virtual func (input, output: Image) -> OpenGLMap { this defaultMap as OpenGLMap }
 	drawQuad: func { this _renderer drawQuad() }
 	drawLines: func (pointList: VectorList<FloatPoint2D>, projection: FloatTransform3D, pen: Pen) {
 		positions := pointList pointer as Float*
-		this _linesShader add("color", pen color normalized)
-		this _linesShader useProgram(null, projection, FloatTransform3D identity)
+		shader := this getLineShader()
+		shader add("color", pen color normalized)
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Should we convert to YUV here?

**Comment by emilwestergren** [comment]
Updated

File: source/draw/gpu/opengl/OpenGLContext.ooc (hunk 2 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Yes, right now it expects red channel to be monochrome. Fix this in a new PR?

File: source/draw/gpu/opengl/shaders/colorToNativeYuv.frag [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#version 300 es
+#extension GL_EXT_YUV_target : enable
+precision highp float;
+uniform vec4 color;
+layout (yuv) out vec4 outColor;
+void main() {
+	outColor = vec4(color.r, 0.5f, 0.5f, 1.0f);
```
Comment (davidpiuva) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: `vec4(color.rgb, 1.0f)`?

File: source/draw/gpu/opengl/OpenGLContext.ooc (hunk 2 above)
Comment (davidpiuva) [inline source/draw/gpu/opengl/OpenGLContext.ooc]: Okay, as long as it actually uses the color in some way.

File: source/draw/gpu/opengl/shaders/colorToNativeYuv.frag (hunk 3 above)
Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: It seems the color is white in RGBA, i.e. (1.0f, 1.0f, 1.0f, 1.0f)

Comment (davidpiuva) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: Next PR then. :)

Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: Feel free to fix it, I don't care as long as it works :)

**Comment by emilwestergren** [comment]
Updated

File: source/draw/gpu/opengl/EGLYuv.ooc [hunk 4]
```diff
@@ -28,5 +28,9 @@ EGLYuv: class extends OpenGLPacked {
 		drawState blendMode = BlendMode Fill
 		super(drawState)
 	}
+	drawLines: override func (pointList: VectorList<FloatPoint2D>, pen: Pen) {
+		yuv := pen color toYuv()
+		super(pointList, Pen new(ColorRgba new(yuv y, yuv u, yuv v, 255), pen width))
```
Comment (emilwestergren) [inline source/draw/gpu/opengl/EGLYuv.ooc]: Updated

**Comment by davidpiuva** [comment]
Looks good

File: source/draw/gpu/opengl/shaders/colorToNativeYuv.frag [hunk 5]
```diff
@@ -0,0 +1,8 @@
+#version 300 es
+#extension GL_EXT_YUV_target : enable
+precision highp float;
+uniform vec4 color;
+layout (yuv) out vec4 outColor;
+void main() {
+	outColor = color;
```
Comment (marcusnaslund) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: Really complex. Nice job.

Comment (emilwestergren) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: Couldn't have done it without you, this is a team effort guys!

Comment (marcusnaslund) [inline source/draw/gpu/opengl/shaders/colorToNativeYuv.frag]: :heart:
